Renames TELEMETRY setting to ANALYTICS.

dkliban opened the pull request:
fixes: #3417

mdellweg commented on pulpcore/app/util.py:
```diff
@@ -247,11 +247,11 @@ def gpg_verify(public_keys, signature, detached_data=None):
     dispatch_interval = timedelta(days=1)
-    name = "Post Anonymous Telemetry Periodically"
-    if settings.TELEMETRY:
+    name = "Post Anonymous Analytics Periodically"
```
This will leave the old schedule lying around. Can you write an elidable migration to delete it?

dkliban replied:
Yes, I'll add a migration.

mdellweg commented on pulpcore/app/settings.py:
```diff
@@ -289,7 +289,7 @@
 TASK_DIAGNOSTICS = False
 
-TELEMETRY = True
+ANALYTICS = True
```
To make this change not breaking today, can you deprecate but still use the old setting (if set)? I'd say (if set) the new setting ought to take precedence.

dkliban replied:
How long will we support the old setting? I  would like to drop it in a future release.

dkliban replied:
I propose remove `TELEMETRY` in 3.23.

dkliban replied:
We are keeping it around until 4.0

mdellweg commented on CHANGES/3417.bugfix:
```diff
@@ -0,0 +1 @@
+Renamed `TELEMETRY` setting to `ANALYTICS`. This is a necessary, but backwards incompatible change.
```
This should come with a removal changelog entry.

dkliban replied:
I'll add one.

ipanova commented on CHANGES/3417.bugfix:
```diff
@@ -0,0 +1 @@
+Renamed `TELEMETRY` setting to `ANALYTICS`. This is a necessary, but backwards incompatible change.
```
Imagine user had set the telemetry to False and not we just replace it with analytics which is enabled by defaults. it means once the user will upgrade to next pulpcore version the data will be collected again, even if he previously marked the setting to False.

dkliban replied:
We don't have a deprecation policy for settings/user features. I propose that we deprecate in 3.22 and remove in 3.23.

ipanova replied:
that's fine with me as long as we don't delete that setting right away, first deprecate it and introduce a new analytics settings, basically this https://github.com/pulp/pulpcore/pull/3423#discussion_r1029522070

dkliban commented:
After some discussion with @mdellweg on Matrix I left the `TELEMETRY` setting in place and introduced a new `ANALYTICS` setting. I also added a migration to remove the old TaskSchedule.

mdellweg commented:
> After some discussion with @mdellweg on Matrix I left the `TELEMETRY` setting in place and introduced a new `ANALYTICS` setting. I also added a migration to remove the old TaskSchedule.

To sum up my contribution to that discussion: Leaving the setting around as an alias forever is a small price payed for not upsetting users.

mdellweg commented on pulpcore/app/migrations/0097_remove_telemetry_task_schedule.py:
```diff
@@ -0,0 +1,25 @@
+
+    operations = [
+        migrations.RunPython(
+            remove_telemetry_task_schedule, reverse_code=migrations.RunPython.noop
```
```suggestion
            remove_telemetry_task_schedule, reverse_code=migrations.RunPython.noop, elidable=True
```

mdellweg reviewed (approved):
One optional suggestion...

mdellweg commented on pulpcore/app/settings.py:
```diff
@@ -289,8 +289,11 @@
 
 TASK_DIAGNOSTICS = False
 
+# Deprecated
 TELEMETRY = True
 
+ANALYTICS = TELEMETRY
+
```
Actually i doubt that this works as desired. After executing this module, `ANALYTICS` will be `True` and not "the value of `TELEMETRY`".

dkliban replied:
I'll try to add a unit test around this.

daviddavis commented:
> To sum up my contribution to that discussion: Leaving the setting around as an alias forever is a small price payed for not upsetting users.

We have to set `TELEMETRY` to false in some of our environments for security reasons so maintaining it as an option until the next major release of pulp would make us happy. Also, my hope would be that settings would be covered by Pulp's semver policy.

Thanks 👍

bmbouter commented:
I can agree to maintain this setting until the next major version change. It's not my first choice, but that's the feedback we got from users, so let's work with that.

Also I think the new setting should be called `PULP_ANALYTICS`. I'm thinking all Pulp specific new settings should be prepended with `PULP_` to allow Pulp to be safely combined with any other random Django project which is a deployment scenario we allow.

mdellweg commented:
> Also I think the new setting should be called `PULP_ANALYTICS`. I'm thinking all Pulp specific new settings should be prepended with `PULP_` to allow Pulp to be safely combined with any other random Django project which is a deployment scenario we allow.

When translating to environment variables, they will automatically be prefixed with "PULP_". In plugins is saw variables being prefixed by the plugin name. I'm kind of against `PULP_PULP_ANALYTICS=true` and I'm not sure if `PULP_CORE_ANALYTICS=true` would carry the meaning we want.

bmbouter commented:
> When translating to environment variables, they will automatically be prefixed with "PULP_". In plugins is saw variables being prefixed by the plugin name. I'm kind of against `PULP_PULP_ANALYTICS=true` and I'm not sure if `PULP_CORE_ANALYTICS=true` would carry the meaning we want.

This is a great point, let's not prepend settings with `PULP_`, having a user specify an env var as `PULP_PULP_ANALYTICS` sounds not good. So given that discussion, Having the original setting name of `ANALYTICS` is what makes the most sense to me.

bmbouter commented on CHANGES/3417.deprecation:
```diff
@@ -0,0 +1 @@
+`TELEMETRY` setting got deprecated in favor of `ANALYTICS`.
```
```suggestion
``TELEMETRY`` setting was deprecated in favor of ``ANALYTICS``.
```

It needs double backticks.

bmbouter commented on CHANGES/3417.feature:
```diff
@@ -0,0 +1 @@
+The new `ANALYTICS` setting replaced the `TELEMETRY` setting to avoid confusion with application metrics that will be exposed using `OpenTelemetry`.
```
```suggestion
The new ``ANALYTICS`` setting replaced the ``TELEMETRY`` setting to avoid confusion with application metrics that will be exposed using Open Telemetry.
```

bmbouter replied:
Also I think this could be line wrapped maybe.

bmbouter reviewed (approved):
I left some small formatting/language requests, but this is looking good. Thanks!

bmbouter commented on pulpcore/app/util.py:
```diff
@@ -247,11 +247,15 @@ def gpg_verify(public_keys, signature, detached_data=None):
-    if settings.TELEMETRY:
+    name = "Post Anonymous Analytics Periodically"
+    analytics = settings.ANALYTICS
+    if settings.get("TELEMETRY") is not None:
```
A 1-liner comment in this if block for our future selves would be nice.

mdellweg reviewed (approved):
Now we confuse our own CI. (The deprecation check still works...)
